In [2]:
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow import keras
import matplotlib.image as mpimg
from collections import Counter

In [3]:
train_data = pd.read_csv('train.csv')
test_data = pd.read_csv('test.csv')

In [4]:
train_data.head()

id keyword location                                               text  \
0   1     NaN      NaN  Our Deeds are the Reason of this #earthquake M...   
1   4     NaN      NaN             Forest fire near La Ronge Sask. Canada   
2   5     NaN      NaN  All residents asked to 'shelter in place' are ...   
3   6     NaN      NaN  13,000 people receive #wildfires evacuation or...   
4   7     NaN      NaN  Just got sent this photo from Ruby #Alaska as ...   

   target  
0       1  
1       1  
2       1  
3       1  
4       1

In [5]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7613 entries, 0 to 7612
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        7613 non-null   int64 
 1   keyword   7552 non-null   object
 2   location  5080 non-null   object
 3   text      7613 non-null   object
 4   target    7613 non-null   int64 
dtypes: int64(2), object(3)
memory usage: 297.5+ KB


Predict whether a given tweet is about a real distaster or not.

In [6]:
tokenizer = keras.preprocessing.text.Tokenizer(char_level=True)

In [7]:
vocabulary = Counter()
for tweet in train_data['text']:
    #tweet = tf.strings.regex_replace(tweet, b'[^a-zA-Z]', b' ')
    tweet = tf.strings.split(tweet)
    vocabulary.update(list(tweet.numpy()))

In [8]:
len(vocabulary.keys())

31924

It looks like we have 31924 words in our dictionary. Lets take the top 10,000 of them and used them in our model.

In [9]:
vocab_size = 10000
truncated_vocabulary = [ word for word, count in vocabulary.most_common()[:vocab_size]]

In [10]:
#preprocess to replace each word with its ID (index)
words = tf.constant(truncated_vocabulary)
word_ids = tf.range(len(truncated_vocabulary), dtype=tf.int64)
vocab_init = tf.lookup.KeyValueTensorInitializer(words, word_ids)
num_oov_buckets = 1000
table = tf.lookup.StaticVocabularyTable(vocab_init,num_oov_buckets)

In [11]:
table.lookup(tf.constant([b'this movie was faaaaaaantastic'.split()]))

<tf.Tensor: shape=(1, 4), dtype=int64, numpy=array([[   24,   337,    21, 10770]], dtype=int64)>

In [12]:
def encode_words(X_batch, y_batch):
    return table.lookup(X_batch), y_batch

In [13]:
train_set = train_data[['text','target']]
target = train_set.pop('target')
target = target.values.reshape((7613, 1))
dataset = tf.data.Dataset.from_tensor_slices((train_set.values, target))
train_set_model1 = dataset.batch(32).map(encode_words).prefetch(1)

Lets make sure that the dataset converted correctly. We are going to iterate over the dataset and print it's features and target

In [14]:
for feat, targ in dataset.take(5):
  print ('Features: {}, Target: {}'.format(feat, targ))

Features: [b'Our Deeds are the Reason of this #earthquake May ALLAH Forgive us all'], Target: [1]
Features: [b'Forest fire near La Ronge Sask. Canada'], Target: [1]
Features: [b"All residents asked to 'shelter in place' are being notified by officers. No other evacuation or shelter in place orders are expected"], Target: [1]
Features: [b'13,000 people receive #wildfires evacuation orders in California '], Target: [1]
Features: [b'Just got sent this photo from Ruby #Alaska as smoke from #wildfires pours into a school '], Target: [1]


In [34]:
embed_size = 128
model = keras.models.Sequential([
    keras.layers.Embedding(vocab_size + num_oov_buckets, embed_size, input_shape=[None]),
    keras.layers.GRU(128, return_sequences=True),
    keras.layers.GRU(128),
    keras.layers.Dense(1, activation='sigmoid')
])
model.compile(loss= 'binary_crossentropy', optimizer='Adam', metrics=['accuracy'])

trained_model1 = model.fit(train_set_model1, epochs=25)

Epoch 1/25
238/238 [==============================] - 4s 17ms/step - loss: 0.6889 - accuracy: 0.5684
Epoch 2/25
238/238 [==============================] - 4s 17ms/step - loss: 0.6623 - accuracy: 0.6087
Epoch 3/25
238/238 [==============================] - 4s 17ms/step - loss: 0.5959 - accuracy: 0.6708
Epoch 4/25
238/238 [==============================] - 4s 17ms/step - loss: 0.5718 - accuracy: 0.6851
Epoch 5/25
238/238 [==============================] - 4s 16ms/step - loss: 0.5615 - accuracy: 0.6909
Epoch 6/25
238/238 [==============================] - 4s 17ms/step - loss: 0.5553 - accuracy: 0.6962
Epoch 7/25
238/238 [==============================] - 4s 17ms/step - loss: 0.5542 - accuracy: 0.6958
Epoch 8/25
238/238 [==============================] - 4s 17ms/step - loss: 0.5510 - accuracy: 0.7033
Epoch 9/25
238/238 [==============================] - 4s 17ms/step - loss: 0.5527 - accuracy: 0.7005
Epoch 10/25
238/238 [==============================] - 4s 17ms/step - loss: 0.5501 - accura

In [17]:
import tensorflow_hub as hub

model = keras.Sequential([
    hub.KerasLayer('https://tfhub.dev/google/tf2-preview/nnlm-en-dim50/1', dtype=tf.string, input_shape=[], output_shape=[50]),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [18]:
# This is broken
batch_size = 32
train_set = dataset.batch(batch_size).prefetch(1)
tweet_model = model.fit(train_set, epochs=5)

Epoch 1/5
238/238 [==============================] - 15s 65ms/step - loss: 0.5659 - accuracy: 0.7344
Epoch 2/5
238/238 [==============================] - 15s 64ms/step - loss: 0.5064 - accuracy: 0.7607
Epoch 3/5
238/238 [==============================] - 15s 63ms/step - loss: 0.4955 - accuracy: 0.7686
Epoch 4/5
238/238 [==============================] - 15s 64ms/step - loss: 0.4893 - accuracy: 0.7729
Epoch 5/5
238/238 [==============================] - 15s 64ms/step - loss: 0.4845 - accuracy: 0.7754
